File: python/cryptology/classical/substitution/monoalphabetic/caesar.py
"""
Caesar Cipher implementation.

The Caesar cipher is one of the simplest and most widely known encryption techniques.
It is a type of substitution cipher where each letter in the plaintext is shifted
a certain number of places down the alphabet.
"""

# Default English alphabet (lowercase only)
DEFAULT_ALPHABET = "abcdefghijklmnopqrstuvwxyz"


def encrypt(plaintext: str, shift: int = 3, alphabet: str = DEFAULT_ALPHABET) -> str:
    """
    Encrypt plaintext using Caesar cipher.
    
    Args:
        plaintext: The text to encrypt (will be converted to lowercase)
        shift: The number of positions to shift (default: 3)
        alphabet: The alphabet to use for encryption (default: English lowercase)
    
    Returns:
        The encrypted ciphertext
    
    Example:
        >>> encrypt("HELLO", 3)
        'khoor'
        >>> encrypt("abc", 1)
        'bcd'
    """
    # Convert input to lowercase
    plaintext = plaintext.lower()
    result = []
    
    for char in plaintext:
        if char in alphabet:
            # Find position in alphabet
            pos = alphabet.index(char)
            # Shift the character
            new_pos = (pos + shift) % len(alphabet)
            result.append(alphabet[new_pos])
        else:
            # Keep characters not in alphabet unchanged
            result.append(char)
    
    return ''.join(result)


def decrypt(ciphertext: str, shift: int = 3, alphabet: str = DEFAULT_ALPHABET) -> str:
    """
    Decrypt ciphertext using Caesar cipher.
    
    Args:
        ciphertext: The text to decrypt (will be converted to lowercase)
        shift: The number of positions that were shifted (default: 3)
        alphabet: The alphabet that was used for encryption (default: English lowercase)
    
    Returns:
        The decrypted plaintext
    
    Example:
        >>> decrypt("KHOOR", 3)
        'hello'
    """
    return encrypt(ciphertext, -shift, alphabet)

File: c/src/classical/substitution/monoalphabetic/caesar.c
/**
 * @file caesar.c
 * @brief Caesar Cipher implementation
 */

#include "cryptology/classical/substitution/monoalphabetic/caesar.h"
#include <string.h>
#include <ctype.h>

#define DEFAULT_ALPHABET "abcdefghijklmnopqrstuvwxyz"

static int find_char_in_alphabet(char c, const char *alphabet) {
    const char *pos = strchr(alphabet, c);
    return pos ? (int)(pos - alphabet) : -1;
}

int caesar_encrypt(const char *plaintext, int shift, const char *alphabet,
                   char *result, size_t result_size) {
    if (!plaintext || !result || result_size == 0) {
        return -1;
    }
    
    if (!alphabet) {
        alphabet = DEFAULT_ALPHABET;
    }
    
    size_t alphabet_len = strlen(alphabet);
    size_t plaintext_len = strlen(plaintext);
    
    if (plaintext_len >= result_size) {
        return -1;
    }
    
    // Normalize shift to be positive
    shift = ((shift % (int)alphabet_len) + (int)alphabet_len) % (int)alphabet_len;
    
    for (size_t i = 0; i < plaintext_len; i++) {
        char c = tolower((unsigned char)plaintext[i]);
        int pos = find_char_in_alphabet(c, alphabet);
        
        if (pos != -1) {
            // Character is in alphabet - shift it
            int new_pos = (pos + shift) % alphabet_len;
            result[i] = alphabet[new_pos];
        } else {
            // Character not in alphabet - keep unchanged
            result[i] = plaintext[i];
        }
    }
    
    result[plaintext_len] = '\0';
    return 0;
}

int caesar_decrypt(const char *ciphertext, int shift, const char *alphabet,
                   char *result, size_t result_size) {
    return caesar_encrypt(ciphertext, -shift, alphabet, result, result_size);
}

int caesar_produce_alphabet(int shift, const char *alphabet, char *result, size_t result_size) {
    if (!alphabet || !result || result_size == 0) {
        return -1;
    }
    
    size_t alphabet_len = strlen(alphabet);
    if (alphabet_len >= result_size) {
        return -1;
    }
    
    shift = shift % (int)alphabet_len;
    if (shift < 0) {
        shift += (int)alphabet_len;
    }
    
    strcpy(result, alphabet + shift);
    strncat(result, alphabet, shift);
    result[result_size - 1] = '\0';
    
    return 0;
}

